### PR TITLE
os/bluestore: use buffer values instead of pointers in buffer_map

### PR DESCRIPTION
**Pull request by pereman2**
buffer_map is a map<int, unique_ptr<Buffer>>, therefore, when reading a
Buffer we will have 2 cache misses. This commit refactors buffer_map to
be a map<int, Buffer> so that Buffers are contigous in memory to improve
data locality and reduce cache misses

**`BufferSpace::read` sequential reads to onode:**
  | Minimum: | Maximum: | Quartile 1: | Median: | Quartile 3:
-- | -- | -- | -- | -- | --
before | 30 | 186227 | 170 | 190 | 300
after | 30 | 84209 | 160 | 170 | 200
  | 0 | +54.78% | +5.88% | +10.53% | +33.33%

**`BufferSpace::read` random reads reads to onode:**
Minimum: | Maximum: | Quartile 1: | Median: | Quartile 3:
-- | -- | -- | -- | --
40 | 39099 | 290 | 320 | 350
40 | 21880 | 290 | 320 | 360
0.00% | +44.04% | 0.00% | 0.00% | -2.86%

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -1849,6 +1842,11 @@ void BlueStore::BufferSpace::read(
   uint64_t miss_bytes = want_bytes - hit_bytes;
   cache->logger->inc(l_bluestore_buffer_hit_bytes, hit_bytes);
   cache->logger->inc(l_bluestore_buffer_miss_bytes, miss_bytes);
+  auto end_time = std::chrono::high_resolution_clock::now();
+  auto duration =
+      std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start);
+  std::cout << "Time taken: " << duration.count() << " ns"
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:1848]: to be removed?

**Review by ifed01 — CHANGES_REQUESTED** [review]
This patch is failing ceph_test_objectstore (backtrace could vary a bit):
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ObjectStore/StoreTest
[ RUN ] ObjectStore/StoreTest.Synthetic/1
seeding object 0
seeding object 500
Op 0
available_objects: 1000 in_flight_objects: 0 total objects: 1000 in_flight 0
Op 1000
available_objects: 999 in_flight_objects: 0 total objects: 999 in_flight 0
Op 2000
available_objects: 975 in_flight_objects: 0 total objects: 975 in_flight 0
*** Caught signal (Segmentation fault) **
in thread 7f0b360caa40 thread_name:ceph_test_objec
ceph version Development (no_version) reef (dev)
1: /lib64/libc.so.6(+0x3f1b0) [0x7f0b3543f1b0]
2: (TwoQBufferCacheShard::_trim_to(unsigned long)+0x22b) [0x557e65d744bb]
3: (BlueStore::_shutdown_cache()+0x84) [0x557e65d199d4]
4: (BlueStore::umount()+0xca) [0x557e65d4b73a]
5: (SyntheticWorkloadState::fsck(bool)+0x93) [0x557e65bdb433]
6: bin/ceph_test_objectstore(+0x461c4d) [0x557e65b8fc4d]
7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::)(), char const)+0x4d) [0x557e65e5c36d]

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -378,56 +380,77 @@ class BlueStore : public ObjectStore,
       ceph_assert(writing.empty());
     }
 
-    void _add_buffer(BufferCacheShard* cache, Buffer* b, int level, Buffer* near) {
+    void _add_buffer(BufferCacheShard *cache, BufferSpace *space, unsigned s,
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:383]: May be better to use ctor with move semantics for Buffer instead to avoid BufferSpace interface growth?
See attached patch for possible implementation.

Comment (ifed01) [reply src/os/bluestore/BlueStore.h:383]: diff --git a/src/os/bluestore/BlueStore.cc b/src/os/bluestore/BlueStore.cc
index e6e1aa0541f..ed8eb34fd92 100644
--- a/src/os/bluestore/BlueStore.cc
+++ b/src/os/bluestore/BlueStore.cc
@@ -1728,9 +1728,9 @@ int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint32_t offset, u
 	if (b->data.length()) {
 	  bufferlist bl;
 	  bl.substr_of(b->data, b->length - tail, tail);
-	  _add_buffer(cache, this, b->state, b->seq, end, bl, b->flags, 0, 0, b);
+	  _add_buffer(cache, std::move(Buffer(this, b->state, b->seq, end, bl, b->flags)), 0, 0, b);
 	} else {
-	  _add_buffer(cache, this, b->state, b->seq, end, tail, b->flags, 0, 0, b);
+	  _add_buffer(cache, std::move(Buffer(this, b->state, b->seq, end, tail, b->flags)), 0, 0, b);
 	}
 	if (!b->is_writing()) {
 	  cache->_adjust_size(b, front - (int64_t)b->length);
@@ -1760,9 +1760,13 @@ int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint32_t offset, u
     if (b->data.length()) {
       bufferlist bl;
       bl.substr_of(b->data, b->length - keep, keep);
-      _add_buffer(cache, this, b->state, b->seq, end, bl, b->flags, 0, 0, b);
+      _add_buffer(cache,
+                  std::move(Buffer(this, b->state, b->seq, end, bl, b->flags)),
+                  0, 0, b);
     } else {
-      _add_buffer(cache, this, b->state, b->seq, end, keep, b->flags, 0, 0, b);
+      _add_buffer(cache,
+                  std::move(Buffer(this, b->state, b->seq, end, keep, b->flags)),
+                  0, 0, b);
     }
     _rm_buffer(cache, i);
     cache->_audit("discard end 2");
@@ -1779,7 +1783,6 @@ void BlueStore::BufferSpace::read(
   interval_set<uint32_t>& res_intervals,
   int flags)
 {
-  auto start = std::chrono::high_resolution_clock::now();
   res.clear();
   res_intervals.clear();
   uint32_t want_bytes = length;
@@ -1842,11 +1845,6 @@ void BlueStore::BufferSpace::read(
   uint64_t miss_bytes = want_bytes - hit_bytes;
   cache->logger->inc(l_bluestore_buffer_hit_bytes, hit_bytes);
   cache->logger->inc(l_bluestore_buffer_miss_bytes, miss_bytes);
-  auto end_time = std::chrono::high_resolution_clock::now();
-  auto duration =
-      std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start);
-  std::cout << "Time taken: " << duration.count() << " ns"
-            << std::endl;
 }
 
 void BlueStore::BufferSpace::_finish_write(BufferCacheShard* cache, uint64_t seq)
@@ -1895,7 +1893,10 @@ bool BlueStore::BufferSpace::_dup_writing(BufferCacheShard* cache, BufferSpace*
     for (auto it = writing.begin(); it != writing.end(); ++it) {
       Buffer& b = *it;
       ceph_assert(b.is_writing());
-      to->_add_buffer(cache, to, b.state, b.seq, b.offset, b.data, b.flags, 0, 0, nullptr);
+      to->_add_buffer(cache,
+                      std::move(
+                        Buffer(to, b.state, b.seq, b.offset, b.data, b.flags)),
+                      0, 0, nullptr);
     }
   }
   return copied;
@@ -1918,11 +1919,15 @@ void BlueStore::BufferSpace::split(BufferCacheShard* cache, size_t pos, BlueStor
       if (p->second.data.length()) {
         bufferlist bl;
         bl.substr_of(p->second.data, left, right);
-        r._add_buffer(cache, &r, p->second.state, p->second.seq, 0, bl,
-                      p->second.flags, 0, 0, &p->second);
+        r._add_buffer(cache,
+                      std::move(Buffer(&r, p->second.state, p->second.seq,
+                                0, bl, p->second.flags)),
+                      0, 0, &p->second);
       } else {
-        r._add_buffer(cache, &r, p->second.state, p->second.seq, 0, right,
-                      p->second.flags, 0, 0, &p->second);
+        r._add_buffer(cache,
+                      std::move(Buffer(&r, p->second.state, p->second.seq,
+                                       0, right, p->second.flags)),
+                      0, 0, &p->second);
       }
       cache->_adjust_size(&p->second, -right);
       p->second.truncate(left);
@@ -1932,12 +1937,16 @@ void BlueStore::BufferSpace::split(BufferCacheShard* cache, size_t pos, BlueStor
     ceph_assert(p->second.end() > pos);
     ldout(cache->cct, 30) << __func__ << " move " << p->second << dendl;
     if (p->second.data.length()) {
-      r._add_buffer(cache, &r, p->second.state, p->second.seq,
-                    p->second.offset - pos, p->second.data, p->second.flags, 0,
-                    0, &p->second);
+      r._add_buffer(cache, std::move(
+                      Buffer(&r, p->second.state, p->second.seq,
+                             p->second.offset - pos, p->second.data,
+                             p->second.flags)),
+                    0, 0, &p->second);
     } else {
-      r._add_buffer(cache, &r, p->second.state, p->second.seq,
-                    p->second.offset - pos, p->second.length, p->second.flags,
+      r._add_buffer(cache, std::move(
+                      Buffer(&r, p->second.state, p->second.seq,
+                             p->second.offset - pos, p->second.length,
+                             p->second.flags)),
                     0, 0, &p->second);
     }
     if (p == buffer_map.begin()) {
@@ -2871,7 +2880,7 @@ uint32_t BlueStore::Blob::merge_blob(CephContext* cct, Blob* blob_to_dissolve)
     auto buf = src->bc.buffer_map.extract(src->bc.buffer_map.cbegin());
     buf.mapped().space = &dst->bc;
     if (dst->bc.buffer_map.count(buf.key()) == 0) {
-      dst->bc.buffer_map.insert({buf.key(), std::move(buf.mapped())});
+      dst->bc.buffer_map.emplace(std::make_pair(buf.key(), std::move(buf.mapped())));
     }
   }
   // move BufferSpace writing
diff --git a/src/os/bluestore/BlueStore.h b/src/os/bluestore/BlueStore.h
index d8c2893e0c1..b1cdf555cd1 100644
--- a/src/os/bluestore/BlueStore.h
+++ b/src/os/bluestore/BlueStore.h
@@ -304,6 +304,19 @@ public:
     boost::intrusive::list_member_hook<> lru_item;
     boost::intrusive::list_member_hook<> state_item;
 
+    Buffer(Buffer&& other) {
+      std::swap(space, other.space);
+      std::swap(state, other.state);
+      std::swap(cache_private, other.cache_private);
+      std::swap(flags, other.flags);
+      std::swap(seq, other.seq);
+      std::swap(offset, other.offset);
+      std::swap(length, other.length);
+      std::swap(data, other.data);
+      std::swap(cache_age_bin, other.cache_age_bin);
+      std::swap(lru_item, other.lru_item);
+      std::swap(state_item, other.state_item);
+    }
     Buffer(BufferSpace *space, unsigned s, uint64_t q, uint32_t o, uint32_t l,
 	   unsigned f = 0)
       : space(space), state(s), flags(f), seq(q), offset(o), length(l) {}
@@ -380,31 +393,13 @@ public:
       ceph_assert(writing.empty());
     }
 
-    void _add_buffer(BufferCacheShard *cache, BufferSpace *space, unsigned s,
-                     uint64_t q, uint32_t o, uint16_t length, unsigned f,
-                     uint16_t cache_private, int level, Buffer *near) {
-      auto it =
-          buffer_map.emplace(std::piecewise_construct, std::forward_as_tuple(o),
-                             std::forward_as_tuple(space, s, q, o, length, f));
-      Buffer *buffer = &it.first->second;
-      buffer->cache_private = cache_private;
-      _add_buffer(cache, space, buffer, level, near);
-    }
+    void _add_buffer(BufferCacheShard *cache, Buffer &&b, uint16_t cache_private, int level, Buffer *near) {
+      cache->_audit("_add_buffer start");
+      auto it = buffer_map.emplace(
+        std::make_pair(b.offset, std::move(b))); // b is invalidated after this point
 
-    void _add_buffer(BufferCacheShard *cache, BufferSpace *space, unsigned s,
-                     uint64_t q, uint32_t o, ceph::buffer::list &b, unsigned f,
-                     uint16_t cache_private, int level, Buffer *near) {
-      auto it =
-          buffer_map.emplace(std::piecewise_construct, std::forward_as_tuple(o),
-                             std::forward_as_tuple(space, s, q, o, b, f));
       Buffer *buffer = &it.first->second;
       buffer->cache_private = cache_private;
-      _add_buffer(cache, space, buffer, level, near);
-    }
-
-    __attribute__((always_inline)) void _add_buffer(BufferCacheShard *cache, BufferSpace *space,
-                     Buffer *buffer, int level, Buffer *near) {
-      cache->_audit("_add_buffer start");
       if (buffer->is_writing()) {
         // we might get already cached data for which resetting mempool is inppropriate
         // hence calling try_assign_to_mempool
@@ -471,18 +466,19 @@ public:
     void write(BufferCacheShard* cache, uint64_t seq, uint32_t offset, ceph::buffer::list& bl,
 	       unsigned flags) {
       std::lock_guard l(cache->lock);
-      Buffer b(this, Buffer::STATE_WRITING, seq, offset, bl,
-			     flags);
       uint16_t cache_private = _discard(cache, offset, bl.length());
-      _add_buffer(cache, this, Buffer::STATE_WRITING, seq, offset, bl, flags, cache_private, (flags & Buffer::FLAG_NOCACHE) ? 0 : 1, nullptr);
+      _add_buffer(cache, std::move(
+                    Buffer(this, Buffer::STATE_WRITING, seq, offset, bl, flags)),
+                  cache_private, !(flags & Buffer::FLAG_NOCACHE), nullptr);
       cache->_trim();
     }
     void _finish_write(BufferCacheShard* cache, uint64_t seq);
     void did_read(BufferCacheShard* cache, uint32_t offset, ceph::buffer::list& bl) {
       std::lock_guard l(cache->lock);
-      Buffer b(this, Buffer::STATE_CLEAN, 0, offset, bl);
       uint16_t cache_private = _discard(cache, offset, bl.length());
-      _add_buffer(cache, this, Buffer::STATE_CLEAN, 0, offset, bl, 0, cache_private, 1, nullptr);
+      _add_buffer(cache, std::move(
+                    Buffer(this, Buffer::STATE_CLEAN, 0, offset, bl)),
+                  cache_private, 1, nullptr);
       cache->_trim();
     }

Comment (pereman2) [reply src/os/bluestore/BlueStore.h:383]: I was playing around https://quick-bench.com/ to understand what would happen with `piecewise_construct` or `_add_buffer(std::move(Buffer(...)));` and I had surprising results:

RlsrIac1wT_YFQhEgI2eTsmFjro

(y axis is ms, higher is slower)

As you can see what I did with piecewise_construct is surprisingly slow compared to just your patch. In addition to move, a reference and then insert instead of emplace is just as fast in this small bench.

[benchmark code](https://godbolt.org/z/szz8j8afc)

In conclusion, I will undoubtedly bring your patch alive

Comment (ifed01) [reply src/os/bluestore/BlueStore.h:383]: Nice tool, wasn't aware of it.
I played a bit and got less optimistic view for insert() when using gcc 13.x on x86 and c++ 20.
movector

Comment (pereman2) [reply src/os/bluestore/BlueStore.h:383]: OaHvA52utZs6z4cArgPLwk2fN0Y

I saw a warning in the compiler that `std::move(Buffer(...))` prevents copy elision optimization so I think I'll go without std::move

**Comment by ifed01** [comment]
> This patch is failing ceph_test_objectstore (backtrace could vary a bit): [==========] Running 1 test from 1 test suite. [----------] Global test environment set-up. [----------] 1 test from ObjectStore/StoreTest [ RUN ] ObjectStore/StoreTest.Synthetic/1 seeding object 0 seeding object 500 Op 0 available_objects: 1000 in_flight_objects: 0 total objects: 1000 in_flight 0 Op 1000 available_objects: 999 in_flight_objects: 0 total objects: 999 in_flight 0 Op 2000 available_objects: 975 in_flight_objects: 0 total objects: 975 in_flight 0 *** Caught signal (Segmentation fault) ** in thread 7f0b360caa40 thread_name:ceph_test_objec ceph version Development (no_version) reef (dev) 1: /lib64/libc.so.6(+0x3f1b0) [0x7f0b3543f1b0] 2: (TwoQBufferCacheShard::_trim_to(unsigned long)+0x22b) [0x557e65d744bb] 3: (BlueStore::_shutdown_cache()+0x84) [0x557e65d199d4] 4: (BlueStore::umount()+0xca) [0x557e65d4b73a] 5: (SyntheticWorkloadState::fsck(bool)+0x93) [0x557e65bdb433] 6: bin/ceph_test_objectstore(+0x461c4d) [0x557e65b8fc4d] 7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::)(), char const)+0x4d) [0x557e65e5c36d]

Is this resolved?

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (markhpc) [reply src/os/bluestore/BlueStore.h]: > Nice tool, wasn't aware of it.

Same!  That's pretty cool!

File: src/os/bluestore/BlueStore.h [hunk 3]
```diff
@@ -448,18 +457,25 @@ class BlueStore : public ObjectStore,
     void write(BufferCacheShard* cache, uint64_t seq, uint32_t offset, ceph::buffer::list& bl,
 	       unsigned flags) {
       std::lock_guard l(cache->lock);
-      Buffer *b = new Buffer(this, Buffer::STATE_WRITING, seq, offset, bl,
+      Buffer b(this, Buffer::STATE_WRITING, seq, offset, bl,
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:460]: no-op, please remove

File: src/os/bluestore/BlueStore.h [hunk 4]
```diff
@@ -448,18 +457,25 @@ class BlueStore : public ObjectStore,
...
+                  nullptr);
       cache->_trim();
     }
     void _finish_write(BufferCacheShard* cache, uint64_t seq);
     void did_read(BufferCacheShard* cache, uint32_t offset, ceph::buffer::list& bl) {
       std::lock_guard l(cache->lock);
-      Buffer *b = new Buffer(this, Buffer::STATE_CLEAN, 0, offset, bl);
-      b->cache_private = _discard(cache, offset, bl.length());
-      _add_buffer(cache, b, 1, nullptr);
+      Buffer b(this, Buffer::STATE_CLEAN, 0, offset, bl);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:473]: same here, not needed

File: src/test/objectstore/store_test.cc [hunk 5]
```diff
@@ -5435,6 +5437,55 @@ TEST_P(StoreTestSpecificAUSize, SyntheticMatrixPreferDeferred) {
...
+    r = queue_transaction(store, ch, std::move(t));
+    ASSERT_EQ(0, r);
+  } 
+  // read a bunch of times
+  uint64_t blocks = 4096;
+  uint64_t read_blocks = 16;
+  uint64_t read_size = 4096*read_blocks;
+  bufferlist bl;
+  vector<uint64_t> to_read;
+  for (int i = 0; i < blocks; i += read_blocks) {
```
Comment (ifed01) [inline src/test/objectstore/store_test.cc:5476]: int vs. unsigned int comparison warning.

**Comment by pereman2** [comment]
@ifed01 I'm afraid I'm leaving this pr for now. The bug you pasted and the one I found are related to changing the state of `list_member_hook`s of Buffer because moving a Buffer will fail on hook destructor because it is still linked. For this thing to work it will require probably inheriting base_member_hook so that we can have `safe_hook` + explicit unlinking at will and I'm not sure it's worth the effort.

**Comment by ifed01** [comment]
> @ifed01 I'm afraid I'm leaving this pr for now. The bug you pasted and the one I found are related to changing the state of `list_member_hook`s of Buffer because moving a Buffer will fail on hook destructor because it is still linked. For this thing to work it will require probably inheriting base_member_hook so that we can have `safe_hook` + explicit unlinking at will and I'm not sure it's worth the effort.

@pereman2 - could you please elaborate a bit more on the issue please?

**Comment by pereman2** [comment]
@ifed01 Merging blobs requires moving Buffers from one BufferSpace to another. Using values intead of pointers requires copying Buffer memory from one container to another even going through the stack. Let's say we go through stack and do:
```c
{ //start of block
  buffer = src.bc.writing.front()
  src.bc.writing.pop_front()

  dst.bc.buffer_map.insert(buffer.o buffer)
  dst.bc.writing.push(buffer)
} // buffer dctor called ~Buffer() -> buffer.state_item has a value buffer.lru_item may have value
```
Since both member hooks: state_item and lru_item, may be set we need to manually unlink and the update both of them accordingly.

**Comment by pereman2** [comment]
@ifed01 added your patch!

File: src/test/objectstore/store_test.cc (hunk 5 above)
Comment (ifed01) [reply src/test/objectstore/store_test.cc:5476]: @pereman2 - please fix this.

**Review by ifed01 — COMMENTED** [review]
Overall looks good. A couple of nits:
1. signed vs. unsigned comparison warning in store_test.cc
2. Remove(?) the commit tagged as "DELETE this commit" or remove the tag.

**Comment by pereman2** [comment]
https://pulpito.ceph.com/pdiazbou-2024-01-23_08:54:25-rados-wip-pereman2-buffer-value-distro-default-smithi/

teuthology failures seem unrelated

**Comment by ifed01** [comment]
jenkins test make check

**Review by ifed01 — APPROVED** [review]

**Comment by ifed01** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test api
